File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/accepted/rgl-n2.py
#!/usr/bin/env pypy

import sys
sys.setrecursionlimit(6000)

n = int(raw_input())
res = 0

cost = [[0 for i in range(n+1)] for j in range(n+1)]
active = [False for i in range(n+1)]
cycle = [False for i in range(n+1)]
root = [-1 if i > 0 else 0 for i in range(n+1)]
p = [0 for i in range(n+1)]

for i in range(1,n+1):
  arr = list(map(int, raw_input().split()))
  p[i] = arr[0]
  res += arr[1]
  for j in range(n+1):
    cost[j][i] = arr[j+2] - arr[1]

def dfs(x):
  if root[i] != -1:         pass
  if active[x]:             root[x] = x
  elif root[p[x]] != -1:    root[x] = root[p[x]]
  else:
    active[x] = True
    root[x] = dfs(p[x])
    if active[root[x]]:
      cycle[x] = True
    active[x] = False
  return root[x]

for i in range(n+1): root[i] = dfs(i)

for i in range(1,n+1):
  if root[i] == i:
    ring = [j for j in range(n+1) if root[j] == i and cycle[j]]
    rcost = min(cost[n][j] for j in ring)
    res += rcost
    for j in ring:
      for k in range(n+1):
        cost[k][j] -= rcost
      cost[i][j] = 0

for i in range(n+1):
  for j in range(n+1):
    if i == p[j] or cycle[i] and root[i] == root[j]:
      cost[i][j] = 0

visited = [False for i in range(n+1)]
for z in range(n+1):
  x = n+2
  for i in range(n+1):
    if not visited[i] and (x == n+2 or cost[0][i] < cost[0][x]):
      x = i
  visited[x] = True
  for i in range(n+1):
    cost[0][i] = min(cost[0][i], cost[0][x] + cost[x][i])

print(res + cost[0][n])

File: contest/ICPC/2018/欧洲/NWERC西北欧/data/fastestspeedrun/submissions/slow_input/rgl_n2.java
import java.util.*;

public class rgl_n2 {
  private static final int MAX_N = 3000;

  boolean[] active = new boolean[MAX_N+1];
  boolean[] cycle = new boolean[MAX_N+1];
  int[] root = new int[MAX_N+1];
  int[] p = new int[MAX_N+1];

  long[][] cost = new long[MAX_N+1][MAX_N+1];

  private int dfs(int x){
    if (active[x]){
      root[x]=x;
    }else if (root[p[x]]!=-1){
      root[x]=root[p[x]];
    }else{
      active[x]=true;
      root[x]=dfs(p[x]);
      if (active[root[x]]) cycle[x]=true;
      active[x]=false;
    }
    return root[x];
  }

  private void man(){
    final Scanner nanananaBatman = new Scanner(System.in);
    final int n = nanananaBatman.nextInt() + 1;
    long res=0;

    for (int i=1; i<n; i++){
      p[i] = nanananaBatman.nextInt();
      long t = nanananaBatman.nextLong();
      for (int j=0; j<n; j++){
        cost[j][i] = nanananaBatman.nextLong() - t;
      }
      res += t;
    }

    for (int i=n; i-->0;)
      root[i] = -1;
    for (int i=n; i-->0;)
      if (root[i]==-1)
        root[i]=dfs(i);

    for (int i=1; i<n; i++)
      if (root[i]==i){
        int x=i;
        for (int j=n; j-->0;)
          if (root[j]==i && cycle[j] && cost[n-1][j]<cost[n-1][x])
            x=j;

        final long rcost=cost[n-1][x]; res+=rcost;
        for (int j=n; j-->0;)
          if (root[j]==i && cycle[j])
            for (int k=n; k-->0;)
              cost[k][j]-=rcost;
      }

    for (int i=n; i-->0;)
      for (int j=n; j-->0;)
        if (cost[i][j] < 0){
          throw new Error("Bad cost "+i+" "+j+": " + cost[i][j]);
        }

    for (int i=n; i-->0;)
      for (int j=n; j-->0;)
        if (i==p[j] || root[i]==root[j] && cycle[i])
          cost[i][j]=0;

    boolean[] visited=new boolean[n];
    for (int z=0; z<n; z++){
      int x=n;
      for (int i=n; i-->0;){
        if (!visited[i] && (x==n || cost[0][i]<cost[0][x])){
          x=i;
        }
      }
      visited[x]=true;
      for (int i=n; i-->0;){
        cost[0][i]=Math.min(cost[0][i],cost[0][x]+cost[x][i]);
      }
    }

    System.out.println(cost[0][n-1]+res);
  }

  public static void main(String[] args) {
    new rgl_n2().man();
  }
}
